src/mon/command_multiplexer.h

```diff
@@ -0,0 +1,312 @@
+    std::ostringstream os;
+    os << "invalid operation type " << static_cast<int>(op_type);
+    throw std::invalid_argument(os.str());
+ }
+};
+
+/* 
+This is a base class that provides a customization point for user functions, and is
+what users will indirectly interact with most of the time.
+
+In the context of your multiplexer deployment, there will be a defined type along the lines of:
+    using my_command_t = ceph::command<my_command_ctx, my_command_result>;
+
+Given that, to add a new function of your own, follow these steps: 
+
+    * derive your new type from my_command_t;
+    * define the tags (strings) you want to respond to by listing them in your
+    tags() member function;
+    * override one or both of the prepare() and preprocess() member functions;
+    * register the type of your command with the registry helper
+
+In practice, a minimal function would look something like this:
+
+    struct my_command : public my_command_t
+    {
+        vector<string> tags() { return { "my_command" }; }
+    };
+
+...next, you need to register the type of your command function. The framework deployment will probably
+provide a place for you to do this, most likely a free function called register_commands(). There
+will be a place for you to add your command type. Below, we've added "my_command":
+
+    void register_commands(my_command_registry& commands)
+    {
+     ceph::register_commands<my_command_registry, 
+        my_command
+     >(commands);
+    }
+
+That will automatically register your function, set up the command bindings, set up multiplexing for
+whatever phase (preprocess, prepare) you've defined, and handle interpreting the results of your function,
+leaving you to concentrate on what you want it to do.
+
+Next, you'll probably want to define member functions for only the phases (ie. prepare, preprocess) that you
+wish to respond to. For example, we can extend my_command with a prepare command like so:
+
+    struct my_command : public my_command_t
+    {
+        vector<string> tags() { return { "my_command" }; }
+
+        my_command_result at_prepare(const my_command_ctx& state, string_view cmd);
+    };
+
+...that's it. 
+
+Your function's exact call is in the "cmd" parameter, and any other state is communicated via the
+"state" parameter (which you may call something shorter if you like); the exact content depends on
+what is available in your particular deployment.
+
+Next, let's consider what to return from our functions. This is defined by your local deployment, there 
+is no fixed type. Most of the time, you'll probably see something like this:
+    using my_cmd_result = std::tuple<my_cmd_status, int>;
+
+Where my_cmd_status is something like:
+
+    enum class my_cmd_status : int { failure = 0, success };
+
+...but, once again, nothing in the mini-framework enforces this. Above, the enum is allowed to degrade
+to int because the functions it wraps expects an int <= 0 on failure, > 0 on success back.
+
```

Comment:
I do hear you there. This is a case where expected<> could be a good choice. It looks like it went through LEWG for TS3 by unanimous consent. I'm a tiny bit reluctant to use it quite yet, but if would be a strong improvement here I'm also game-- I think it could be worth prototyping.